zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1073,10 +1073,24 @@ protected void processStateChanged(WatchedEvent event) {
 
     fireStateChangedEvent(event.getState());
 
+    // This following is the case for ZkClient no managing connection
     if (!isManagingZkConnection()) {
+      /*
+       * For SharedZkClient, we will not invoke fireNewSessionEvents but will invoke fireAllEvents
+       * This keeps the same behavior as original SharedZkClient.
+       * Invoking fireAllEvents is important to make sure HelixPropertyStore working correctly.
+       */
+      if ((event.getState() == KeeperState.SyncConnected) && (!_isNewSessionEventFired) && (!"0"
+          .equals(getHexSessionId()))) {
+        fireAllEvents();
+        _isNewSessionEventFired = true;
+      } else if (event.getState() == KeeperState.Expired) {
+        _isNewSessionEventFired = false;
+      }
       return;
     }
 
+    // The following is the case for ZkClient managing the connection
```

Comment:
This comment is a bit confusing.  You mean "The following only applies to dedicated ZkClient in which it manages ZkConnection directly", right?

Reply:
I guess it can be simplified to:
// else, if managing the connection

Anyway, if we do what Huizhi suggested, it should be clearer.

Reply:
@lei-xia, right. Your description is more readable. But @jiajunwang and @hulu like the more concise way. Unless you like this separation of two branch way, I will adopt the more concise way. And add another comment about the goal fo what we will achieve. (Intention)

So later if people refactor this code. They know the intention at lease.